api/src/main/java/org/apache/iceberg/metrics/FixedReservoirHistogram.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.metrics;
+
+import java.util.Arrays;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+/**
+ * A {@link Histogram} implementation that uses a fixed size reservoir like circular buffer. Oldest
+ * observation is replaced by latest observation.
```

Comment:
Okay, after reading this comment I understand what you were talking about more clearly. I was wondering why the sampling would be biased toward the last collected values, but I see that it's because those are the values that you're collecting.

What I would expect from a reservoir algorithm is to choose the final set of values using reservoir sample so that we know that $k$ values from the entire input set, size $n$, are selected with a uniform probability, $k/n$, using the [simple sampling algorithm](https://en.wikipedia.org/wiki/Reservoir_sampling).

I think that we probably want to use the uniform sampling by default. That way it isn't biased toward the order of values. That should also make the choice of reservoir size easier, since it needs to be proportional to accuracy and the target number of ntiles, rather than proportional to the size of the input data.

Reply:
I see where the difference comes from. I was mostly thinking about the long-running streaming job, where metrics are periodically polled and reported to metrics backend. Hence we only want to report the measurements based on latest values.

You were talking about the batch jobs where uniform sampling is more preferable without bias toward the recent values.

Reply:
Are we going to have cases where we want to use these metrics in long-running streaming jobs? Wouldn't we use Flink's metrics directly for that?

What is a case where we want to use `MetricsContext` and have Iceberg report metrics where we'd want to bias toward the latest values?